File: packages/sea-react-native/android/src/main/java/com/bankerise/seareactnative/SEABridgePresenter.kt
package com.bankerise.seareactnative

import android.app.Activity
import android.net.Uri
import android.util.Log
import com.bankerise.sea.core.SEAConfig
import com.bankerise.sea.core.SEAAppearance
import com.bankerise.sea.core.SEAAuthMode
import com.bankerise.sea.core.SEAError
import com.bankerise.sea.core.SEAPresentation
import com.bankerise.sea.core.SEAPropertiesLoader
import com.bankerise.sea.core.SEASession
import org.json.JSONObject

/**
 * The entire bridge-side "logic": marshal primitive prop values into a
 * [SEAConfig], call [SEASession.start], and marshal the result back to
 * primitives. No authentication/navigation/validation logic lives here —
 * all of that is `sea-core-android` (spec §4.2, §7).
 *
 * Only file in this module that imports `com.bankerise.sea.core.*`.
 * Mirrors `ios/SEABridgePresenter.swift` exactly.
 */
object SEABridgePresenter {

    /**
     * Marshals the per-session/UI prop values into [SEAConfig] and
     * [SEAAppearance], calls [SEASession.start], and returns `true` if
     * the auth surface was launched. On validation failure,
     * [callbacks.onError] fires before this returns `false`.
     *
     * Security knobs (callback scheme, allowed domains, allowed ports,
     * max URL length, timeout) are NOT props — they are owned by the
     * platform config: `bankerise-sea.properties` loaded via
     * [SEAPropertiesLoader]. Only `authorizeUrl`, `presentation`,
     * `authMode` and `appearance` cross the bridge. Mirrors
     * `ios/SEABridgePresenter.swift`.
     */
    fun start(
        activity: Activity,
        authorizeUrl: String,
        presentation: String,
        authMode: String,
        headerBackground: Int?,
        headerText: Int?,
        accent: Int?,
        closeIconTint: Int?,
        cornerRadius: Float?,
        title: String,
        callbacks: BridgeCallbacks
    ): Boolean {
        val url = try {
            Uri.parse(authorizeUrl)
        } catch (_: Exception) {
            callbacks.onError("invalid_authorize_url", "malformed authorizeUrl")
            return false
        }

        var appearance = SEAAppearance.default
        if (headerBackground != null) appearance = appearance.copy(headerBackground = headerBackground)
        if (headerText != null) appearance = appearance.copy(headerText = headerText)
        if (accent != null) appearance = appearance.copy(accent = accent)
        if (closeIconTint != null) appearance = appearance.copy(closeIconTint = closeIconTint)
        if (cornerRadius != null && cornerRadius > 0f) appearance = appearance.copy(cornerRadius = cornerRadius)
        if (title.isNotEmpty()) appearance = appearance.copy(title = title)

        // Source of truth for all security knobs (§4.5, §7.1/§7.3): the
        // host app's bankerise-sea.properties. Only the per-session/UI
        // values are overridden here.
        val config = SEAPropertiesLoader.loadConfig(activity).copy(
            authorizeUrl = url,
            presentation = if (presentation == "fullscreen") {
                SEAPresentation.FULLSCREEN
            } else {
                SEAPresentation.SHEET
            },
            appearance = appearance,
            authMode = if (authMode == "nativeBrowser") {
                SEAAuthMode.NATIVE_BROWSER
            } else {
                SEAAuthMode.EMBEDDED
            }
        )

        val seaCallbacks = SEASession.Callbacks(
            onCaptured = { params ->
                val json = jsonString(params.raw)
                Log.d("SEABridgePresenter", "onCaptured: raw=${params.raw} json=$json")
                callbacks.onCaptured(json)
            },
            onCancelled = {
                Log.d("SEABridgePresenter", "onCancelled")
                callbacks.onCancelled()
            },
            onError = { error ->
                val (code, message) = taxonomy(error)
                Log.d("SEABridgePresenter", "onError: code=$code message=$message")
                callbacks.onError(code, message)
            }
        )

        return SEASession.start(activity, config, seaCallbacks)
    }

    /** JSON-serialize a `Map<String, String>` via `org.json.JSONObject`. */
    private fun jsonString(raw: Map<String, String>): String {
        return try {
            val obj = JSONObject()
            raw.forEach { (k, v) -> obj.put(k, v) }
            obj.toString()
        } catch (_: Exception) {
            "{}"
        }
    }

    /**
     * Maps every [SEAError] case to the bridge's wire taxonomy verbatim
     * (spec §7.2): network, timeout, cancelled, invalid_authorize_url,
     * server_error, webauthn_unavailable, kill_switched.
     */
    private fun taxonomy(error: SEAError): Pair<String, String?> = when (error) {
        is SEAError.Network -> "network" to error.underlying
        is SEAError.Timeout -> "timeout" to null
        is SEAError.Cancelled -> "cancelled" to null
        is SEAError.InvalidAuthorizeUrl -> "invalid_authorize_url" to error.reason.name
        is SEAError.ServerError -> "server_error" to error.statusCode.toString()
        is SEAError.WebauthnUnavailable -> "webauthn_unavailable" to null
        is SEAError.KillSwitched -> "kill_switched" to null
    }

    /**
     * Callbacks carrying only primitive/bridgeable types — `sea-core-android`'s
     * own types ([SEACallbackParams][com.bankerise.sea.core.SEACallbackParams],
     * [SEAError]) never cross into the View layer.
     *
     * Mirrors `ios/SEABridgePresenter.swift` `SEABridgeCallbacks`.
     */
    class BridgeCallbacks(
        val onCaptured: (String) -> Unit,
        val onCancelled: () -> Unit,
        val onError: (String, String?) -> Unit
    )
}

File: packages/sea-react-native/ios/SEABridgePresenter.swift
import Foundation
import UIKit
import SEACore

/// `@objc` callback surface for `SeaReactNativeView.mm`. Every closure here
/// carries only primitive/bridgeable types — `SEACore`'s own types
/// (`SEACallbackParams`, `SEAError`) never cross into Objective-C++.
@objc(SEABridgeCallbacks)
public final class SEABridgeCallbacks: NSObject {
    let onCaptured: (String) -> Void
    let onCancelled: () -> Void
    let onError: (String, String?) -> Void

    @objc public init(
        onCaptured: @escaping (String) -> Void,
        onCancelled: @escaping () -> Void,
        onError: @escaping (String, String?) -> Void
    ) {
        self.onCaptured = onCaptured
        self.onCancelled = onCancelled
        self.onError = onError
    }
}

/// The entire bridge-side "logic": marshal primitive prop values into a
/// `SEAConfig`, call `SEASession.start`, and marshal the result back to
/// primitives. No authentication/navigation/validation logic lives here —
/// all of that is `SEACore`'s (spec §4.2, §7).
@objc(SEABridgePresenter)
public final class SEABridgePresenter: NSObject {
    @discardableResult
    @objc public static func start(
        fromAnchor anchorView: UIView,
        authorizeUrl: String,
        presentation: String,
        authMode: String,
        headerBackground: UIColor?,
        headerText: UIColor?,
        accent: UIColor?,
        closeIconTint: UIColor?,
        cornerRadius: NSNumber?,
        title: String,
        callbacks: SEABridgeCallbacks
    ) -> UIViewController? {
        guard let url = URL(string: authorizeUrl) else {
            callbacks.onError("invalid_authorize_url", "malformed authorizeUrl")
            return nil
        }
        guard let presenter = topMostViewController(from: anchorView) else {
            callbacks.onError("cancelled", "no presenter view controller available")
            return nil
        }

        var appearance = SEAAppearance.default
        if let headerBackground { appearance.headerBackground = headerBackground }
        if let headerText { appearance.headerText = headerText }
        if let accent { appearance.accent = accent }
        if let closeIconTint { appearance.closeIconTint = closeIconTint }
        if let cornerRadius { appearance.cornerRadius = CGFloat(truncating: cornerRadius) }
        appearance.title = title.isEmpty ? nil : title

        // Security knobs (allowedDomains, timeoutMs, callbackScheme, ports,
        // URL-length cap) are not bridge props: they are owned by the
        // platform config — `SEASecurityConfig.plist` read by
        // `SEAEnvironment.current` (§3.3) plus `SEAConfig`'s own defaults.
        // Host narrowing and timeout no longer apply from JS (spec §7.1/§7.3).
        let config = SEAConfig(
            authorizeURL: url,
            presentation: presentation == "fullscreen" ? .fullscreen : .sheet,
            appearance: appearance,
            authMode: authMode == "nativeBrowser" ? .nativeBrowser : .embedded
        )

        let seaCallbacks = SEASession.Callbacks(
            onCaptured: { params in
                callbacks.onCaptured(jsonString(from: params.raw))
            },
            onCancelled: {
                callbacks.onCancelled()
            },
            onError: { error in
                let (code, message) = taxonomy(for: error)
                callbacks.onError(code, message)
            }
        )

        return SEASession.start(config: config, from: presenter, callbacks: seaCallbacks)
    }

    @objc public static func dismiss(_ viewController: UIViewController) {
        viewController.dismiss(animated: true)
    }

    private static func jsonString(from raw: [String: String]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: raw) else { return "{}" }
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    /// Maps every `SEAError` case to the bridge's wire taxonomy verbatim
    /// (spec §7.2): network, timeout, cancelled, invalid_authorize_url,
    /// server_error, webauthn_unavailable, kill_switched.
    private static func taxonomy(for error: SEAError) -> (code: String, message: String?) {
        switch error {
        case .network(let underlying):
            return ("network", underlying)
        case .timeout:
            return ("timeout", nil)
        case .cancelled:
            return ("cancelled", nil)
        case .invalidAuthorizeURL(let reason):
            return ("invalid_authorize_url", reason.rawValue)
        case .serverError(let statusCode):
            return ("server_error", "\(statusCode)")
        case .webauthnUnavailable:
            return ("webauthn_unavailable", nil)
        case .killSwitched:
            return ("kill_switched", nil)
        }
    }

    private static func topMostViewController(from view: UIView) -> UIViewController? {
        guard var top = view.window?.rootViewController else { return nil }
        while let presented = top.presentedViewController {
            top = presented
        }
        return top
    }
}
